src/box/sql/where.c

```diff
@@ -2926,6 +2926,9 @@ whereLoopAddBtree(WhereLoopBuilder * pBuilder,	/* WHERE clause information */
 	for (uint32_t i = 0; i < idx_count; iSortIdx++, i++) {
 		if (i > 0)
 			probe = space->index[i]->def;
+		/* Such index may possibly contain not all tuples, so skip it */
+		if (pSrc->pIBIndex == NULL && probe->key_def->has_exclude_null)
```

Comment:
This makes indexes with exclude_null always invisible in SQL, although they can be used. For instance (schema is taken from your test):
```
explain query plan select s2 from j where s2 == 3;
[0, 0, 0, 'SCAN TABLE J (~262144 rows)']
```
If index didn't contain exclude_null option, plan would be following:
`
[0, 0, 0, 'SEARCH TABLE J USING COVERING INDEX I3 (S2=?) (~1 row)']
`
I guess we should come for a bit more sophisticated heuristics (probably it should be implemented not in scope of this issue). Please, find out if we can fix this easily and in case we can't - file an issue.

Reply:
In any case, it is better to do in a new issue - I created it [here](https://github.com/tarantool/tarantool/issues/5718#issue-787979259)